File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = {}
        for x in nums:
            if x not in freq:
                freq[x] = 0
            freq[x] += 1
        # print(freq)
        heap = []
        heapq.heapify(heap)
        for key in freq:
            heapq.heappush(heap, (freq[key], key))
            if len(heap) > k:
                heapq.heappop(heap)
        # print(heap)
        return [tup[1] for tup in heap]
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int cap = ((int)Math.pow(10,4)*2)+2;
        int offset = (int)Math.pow(10,4);
        int[] hash = new int[cap];
        for (int x : nums)
        {
            hash[x+offset]++;
        }
        Queue<Node> heap = new PriorityQueue<>((a,b)->(a.freq-b.freq));
        for (int i=0; i<cap; i++)
        {
            if (hash[i]>0)
            {
                heap.offer(new Node(i-offset, hash[i]));
            }
            if (heap.size()>k)
            {
                heap.poll();
            }
        }
        int resultSize = Math.max(k, heap.size());
        int[] result = new int[resultSize];
        for (int i=0; i<resultSize; i++)
        {
            result[i] = heap.poll().number;
        }
        return result;
    }
}

class Node
{
    int number;
    int freq;

    Node(int number, int freq)
    {
        this.number = number;
        this.freq = freq;
    }
}